src/cider/nrepl/middleware/debug.clj

```diff
@@ -98,30 +98,39 @@
 
 (defn skip-breaks?
   "True if the breakpoint at coordinates should be skipped.
-  If the mode of `*skip-breaks*` is :all, return true.
-
-  Otherwise, the mode should be :deeper, :before, or :trace.
-  If :deeper, return true if the given coordinates are deeper than the
-  coordinates stored in `*skip-breaks*`.
-  If :before, return true if they represent a place before the coordinates
-  in `*skip-breaks*`.
-  If :trace, return false."
+
+  The `*skip-breaks*` map stores a `mode`, `coordinates`, the `code` that it
+  applies to, and a `force?` flag. Behaviour depends on the `mode`:
+   - :all - return true, skipping all breaks
+   - :trace - return false, skip nothing
+   - :deeper - return true if the given coordinates are deeper than the
+               coordinates stored in `*skip-breaks*`, in the same code
+   - :before - return true if the given coordinates represent a place before
+               the coordinates in `*skip-breaks*`, in the same code
+
+  For :deeper and :before, if we are not in the same code (i.e. we have stepped
+  into another instrumented function), then return the value of `force?`.
+  "
   [coordinates]
   (if (seq coordinates)
-    (when-let [{mode :mode skip-coords :coor code :code} @*skip-breaks*]
-      (case mode
-        ;; From :continue, skip everything.
-        :all    true
-        ;; From :trace, never skip.
-        :trace  false
-        ;; From :out, skip some breaks.
-        :deeper (let [parent (take (count skip-coords) coordinates)]
-                  (and (= code (:code *extras*))
-                       (seq= skip-coords parent)
-                       (> (count coordinates) (count parent))))
-        ;; From :here, skip some breaks.
-        :before (and (= code (:code *extras*))
-                     (coord< coordinates skip-coords))))
+    (when-let [{mode :mode skip-coords :coor
+                code-to-skip :code force? :force?} @*skip-breaks*]
+      (let [same-defn? (= code-to-skip (:code *extras*))]
+        (case mode
+          ;; From :continue, skip everything.
+          :all    true
+          ;; From :trace, never skip.
+          :trace  false
+          ;; From :out, skip some breaks.
+          :deeper (or (and force? (not same-defn?))
```

Comment:
Done, I rearranged the logic as you suggest - it looks much clearer now, thanks.